File: frontend/src/types/novel.ts
export type ChapterStatus = 'draft' | 'reviewed' | 'locked'

export interface NovelCreate {
  title: string
  description?: string | null
  genre?: string | null
  style_guide?: string | null
}

export interface NovelUpdate {
  title?: string | null
  description?: string | null
  genre?: string | null
  style_guide?: string | null
}

export interface NovelListItem {
  id: number
  title: string
  description?: string | null
  genre?: string | null
  style_guide?: string | null
  created_at: string
  updated_at: string
}

export interface ChapterCreate {
  title: string
  content?: string
  summary?: string
  status?: ChapterStatus
}

export interface ChapterUpdate {
  title?: string | null
  content?: string | null
  summary?: string | null
  status?: ChapterStatus | null
}

export interface ChapterOut {
  id: number
  novel_id: number
  title: string
  content: string
  summary: string
  status: ChapterStatus
  created_at: string
  updated_at: string
}

export interface NovelOut extends NovelListItem {
  chapters?: ChapterOut[]
}

File: backend/app/schemas/novel.py
import datetime
from typing import List, Literal, Optional

from pydantic import BaseModel, field_validator

ChapterStatus = Literal["draft", "reviewed", "locked"]


class NovelCreate(BaseModel):
    title: str
    description: Optional[str] = None
    genre: Optional[str] = None
    style_guide: Optional[str] = None

    @field_validator("title")
    @classmethod
    def validate_title(cls, v: str) -> str:
        if len(v) > 200:
            raise ValueError("标题长度不能超过 200 个字符")
        return v

    @field_validator("genre")
    @classmethod
    def validate_genre(cls, v: Optional[str]) -> Optional[str]:
        if v is not None and len(v) > 100:
            raise ValueError("类型长度不能超过 100 个字符")
        return v


class NovelUpdate(BaseModel):
    title: Optional[str] = None
    description: Optional[str] = None
    genre: Optional[str] = None
    style_guide: Optional[str] = None

    @field_validator("title")
    @classmethod
    def validate_title(cls, v: Optional[str]) -> Optional[str]:
        if v is not None and len(v) > 200:
            raise ValueError("标题长度不能超过 200 个字符")
        return v

    @field_validator("genre")
    @classmethod
    def validate_genre(cls, v: Optional[str]) -> Optional[str]:
        if v is not None and len(v) > 100:
            raise ValueError("类型长度不能超过 100 个字符")
        return v


class ChapterCreate(BaseModel):
    title: str
    content: str = ""
    summary: str = ""
    status: ChapterStatus = "draft"

    @field_validator("title")
    @classmethod
    def validate_title(cls, v: str) -> str:
        if len(v) > 200:
            raise ValueError("标题长度不能超过 200 个字符")
        return v


class ChapterUpdate(BaseModel):
    title: Optional[str] = None
    content: Optional[str] = None
    summary: Optional[str] = None
    status: Optional[ChapterStatus] = None


class ChapterOut(BaseModel):
    id: int
    novel_id: int
    title: str
    content: str
    summary: str
    status: str
    created_at: datetime.datetime
    updated_at: datetime.datetime

    model_config = {"from_attributes": True}


class NovelOut(BaseModel):
    id: int
    title: str
    description: Optional[str]
    genre: Optional[str]
    style_guide: Optional[str]
    created_at: datetime.datetime
    updated_at: datetime.datetime
    chapters: Optional[List[ChapterOut]] = None

    model_config = {"from_attributes": True}


class NovelListItem(BaseModel):
    id: int
    title: str
    description: Optional[str]
    genre: Optional[str]
    style_guide: Optional[str]
    created_at: datetime.datetime
    updated_at: datetime.datetime

    model_config = {"from_attributes": True}
